{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ea99190f",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages(library(data.table))\n",
    "suppressPackageStartupMessages(library(dplyr))\n",
    "set.seed(43)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "85838410",
   "metadata": {},
   "outputs": [],
   "source": [
    "t2dggi_gws <- fread(\"/lustre/groups/itg/teams/zeggini/users/archit.singh/lambda/summary_statistics/T2D/T2DGGI/GWS_single_gc_t2dggi.txt\", header = TRUE)\n",
    "diamante18_gws <- fread(\"/lustre/groups/itg/teams/zeggini/users/archit.singh/lambda/summary_statistics/T2D/Mahajan.NatGenet2018b.T2D.European.undajusted.BMI/GWS_single_gc_DIAMANTE18.txt\", header = TRUE)\n",
    "diamante22_gws <- fread(\"/lustre/groups/itg/teams/zeggini/users/archit.singh/lambda/summary_statistics/T2D/Mahajan.NatGen2022.DIAMANTE-EUR.sumstat/GWS_single_gc_DIAMANTE22.txt\", header = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "688035e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#renaming appropriate columns\n",
    "diamante18_gws <- diamante18_gws %>% rename(beta = Beta)\n",
    "dim(diamante18_gws)\n",
    "colnames(diamante18_gws)[colnames(diamante18_gws) == \"Chr\"] <- \"CHR\"\n",
    "\n",
    "dim(diamante22_gws)\n",
    "colnames(diamante22_gws)[colnames(diamante22_gws) == \"chromosome(b37)\"] <- \"CHR\"\n",
    "diamante22_gws <- diamante22_gws %>% rename(beta = `Fixed-effects_beta`)\n",
    "\n",
    "dim(t2dggi_gws)\n",
    "t2dggi_gws$CHR <- sub(\"^chr(\\\\d+):.*$\", \"\\\\1\", t2dggi_gws$MarkerName)\n",
    "t2dggi_gws <- t2dggi_gws %>% rename(beta = Effect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdfbb722",
   "metadata": {},
   "outputs": [],
   "source": [
    "calculate_replication <- function(df1, df2, num_simulations = 100000, alpha = 0.0023) {\n",
    "  # Check if necessary columns are present\n",
    "  if (!all(c(\"CHR\", \"id\", \"beta\") %in% colnames(df1)) || !all(c(\"CHR\", \"id\", \"beta\") %in% colnames(df2))) {\n",
    "    stop(\"Both dataframes must contain 'CHR', 'id', and 'beta' columns.\")\n",
    "  }\n",
    "  \n",
    "  # Initialize a results list to store replication averages for each chromosome\n",
    "  replication_results <- list()\n",
    "  \n",
    "  # Loop through each chromosome in df1\n",
    "  for (chr in unique(df1$CHR)) {\n",
    "    # SNPs on the current chromosome\n",
    "    df1_chr <- df1[df1$CHR == chr, ]\n",
    "    \n",
    "    # Number of SNPs on this chromosome\n",
    "    x <- nrow(df1_chr)\n",
    "    \n",
    "    # Calculate the number of replicated SNPs in df2\n",
    "    replicated_snps <- sum(\n",
    "      df1_chr$id %in% df2$id &\n",
    "        sign(df1_chr$beta[df1_chr$id %in% df2$id]) == sign(df2$beta[match(df1_chr$id[df1_chr$id %in% df2$id], df2$id)])\n",
    "    )\n",
    "    \n",
    "    # Calculate the proportion of replication\n",
    "    replication_proportion <- replicated_snps / x\n",
    "    \n",
    "    # Subset df1 without the current chromosome (LOCO approach)\n",
    "    df1_loco <- df1[df1$CHR != chr, ]\n",
    "    \n",
    "    # Initialize a vector to store results of the simulations\n",
    "    simulation_results <- numeric(num_simulations)\n",
    "    \n",
    "    # Perform simulations\n",
    "    for (i in seq_len(num_simulations)) {\n",
    "      # Randomly sample 'x' SNPs from the LOCO data\n",
    "      sampled_snps <- df1_loco[sample(nrow(df1_loco), x, replace = FALSE), ]\n",
    "      \n",
    "      # Count the number of sampled SNPs replicated in df2\n",
    "      simulation_results[i] <- sum(\n",
    "        sampled_snps$id %in% df2$id &\n",
    "          sign(sampled_snps$beta[sampled_snps$id %in% df2$id]) ==\n",
    "          sign(df2$beta[match(sampled_snps$id[sampled_snps$id %in% df2$id], df2$id)])\n",
    "      )\n",
    "    }\n",
    "    \n",
    "    # Calculate the average replication from the simulations\n",
    "    average_simulation <- mean(simulation_results)\n",
    "    \n",
    "    # Calculate a two-tailed p-value and ensure it's bounded between 0 and 1\n",
    "    #p_value <- 2 * min(\n",
    "    # mean(simulation_results >= replicated_snps), # Probability of observing replicated_snps or more\n",
    "    #mean(simulation_results <= replicated_snps)  # Probability of observing replicated_snps or fewer\n",
    "    #)\n",
    "    #p_value <- max(0, min(1, p_value)) # Bound p-value between 0 and 1\n",
    "    p_value_higher <- mean(simulation_results >= replicated_snps)\n",
    "    p_value_lower <- 1-p_value_higher  \n",
    "    final_p_value <- min(p_value_lower, p_value_higher)\n",
    "    # Determine if the actual replication rate is significantly different from the simulations\n",
    "    #significant_lower <- ifelse(p_value_lower < alpha, TRUE, FALSE)\n",
    "    #significant_higher <- ifelse(p_value_higher < alpha, TRUE, FALSE)\n",
    "    \n",
    "    # Store the results\n",
    "    replication_results[[chr]] <- list(\n",
    "      chromosome = chr,\n",
    "      replicated_snps = replicated_snps,\n",
    "      average_simulation = average_simulation,\n",
    "      #p_value = p_value,\n",
    "      p_value_higher = p_value_higher,\n",
    "      p_value_lower = p_value_lower,\n",
    "      final_p_value = final_p_value,\n",
    "      #significant_lower = significant_lower,\n",
    "      #significant_higher = significant_higher,\n",
    "      replication_proportion = replication_proportion\n",
    "    )\n",
    "  }\n",
    "  \n",
    "  # Convert results to a data frame\n",
    "  results_df <- do.call(rbind, lapply(replication_results, as.data.frame))\n",
    "  return(results_df)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3846f52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Assessing rate of replication before correction\n",
    "results_diamante22_t2dggi <- calculate_replication(diamante22_gws, t2dggi_gws)\n",
    "print(results_diamante22_t2dggi)\n",
    "\n",
    "results_diamante18_t2dggi <- calculate_replication(diamante18_gws, t2dggi_gws)\n",
    "print(results_diamante18_t2dggi)\n",
    "\n",
    "results_diamante18_diamante22 <- calculate_replication(diamante18_gws, diamante22_gws)\n",
    "print(results_diamante18_diamante22)\n",
    "\n",
    "#writing results to file\n",
    "#fwrite(results_diamante18_diamante22, \"C:/Users/ArchitSingh/Desktop/lambda_project/data_from_lustre/analysis_results/rate_of_replication/diamante18-diamante22.txt\", sep = \"\\t\")\n",
    "#fwrite(results_diamante18_t2dggi, \"C:/Users/ArchitSingh/Desktop/lambda_project/data_from_lustre/analysis_results/rate_of_replication/diamante18-t2dggi.txt\", sep = \"\\t\")\n",
    "#fwrite(results_diamante22_t2dggi, \"C:/Users/ArchitSingh/Desktop/lambda_project/data_from_lustre/analysis_results/rate_of_replication/diamante22-t2dggi.txt\", sep = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "606e8501",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Assessing rate of replication after GC correction\n",
    "#correcting the p-values with the new lambda values; check jupyter notebook titled \"getting_variants_with_MAF_morethan_0.005\"\n",
    "#DIAMANTE-18\n",
    "chisq <- qchisq(as.numeric(diamante18_gws$single_gc_p_values),1,lower.tail=FALSE)\n",
    "lambda <- c(1.387) #rounding off to 3 digits after decimal point\n",
    "newchisq <- chisq/lambda\n",
    "newpvalues <- pchisq(newchisq, df=1,lower.tail=FALSE)\n",
    "diamante18_gws$double_gc_p_values <- newpvalues\n",
    "\n",
    "#DIAMANTE-22\n",
    "chisq <- qchisq(as.numeric(diamante22_gws$single_gc_p_values),1,lower.tail=FALSE)\n",
    "lambda <- c(1.402) \n",
    "newchisq <- chisq/lambda\n",
    "newpvalues <- pchisq(newchisq, df=1,lower.tail=FALSE)\n",
    "diamante22_gws$double_gc_p_values <- newpvalues\n",
    "\n",
    "#identifying variants significant after GC correction\n",
    "diamante18_gws_GC <- subset(diamante18_gws, diamante18_gws$double_gc_p_values <= 0.00000005)\n",
    "diamante22_gws_GC <- subset(diamante22_gws, diamante22_gws$double_gc_p_values <= 0.00000005)\n",
    "\n",
    "#running simulation\n",
    "results_diamante22_t2dggi_GC <- calculate_replication(diamante22_gws_GC, t2dggi_gws)\n",
    "print(results_diamante22_t2dggi_GC)\n",
    "\n",
    "results_diamante18_t2dggi_GC <- calculate_replication(diamante18_gws_GC, t2dggi_gws)\n",
    "print(results_diamante18_t2dggi_GC)\n",
    "\n",
    "results_diamante18_diamante22_GC <- calculate_replication(diamante18_gws_GC, diamante22_gws)\n",
    "print(results_diamante18_diamante22_GC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c2e07d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Assessing rate of replication after LDSR intercept correction\n",
    "#Correcting the uncorrected P-values using the LDSR-intercept\n",
    "intercept1 <- c(1.097)\n",
    "chisq <- qchisq(as.numeric(diamante18_gws$single_gc_p_values),1,lower.tail=FALSE)\n",
    "newchisq_intercept <- chisq/intercept1\n",
    "newpvalues <- pchisq(newchisq_intercept, df=1,lower.tail=FALSE)\n",
    "diamante18_gws$intercept_adj_p_values <- newpvalues\n",
    "\n",
    "intercept2 <- c(1.0918)\n",
    "chisq <- qchisq(as.numeric(diamante22_gws$single_gc_p_values),1,lower.tail=FALSE)\n",
    "newchisq_intercept <- chisq/intercept2\n",
    "newpvalues <- pchisq(newchisq_intercept, df=1,lower.tail=FALSE)\n",
    "diamante22_gws$intercept_adj_p_values <- newpvalues\n",
    "\n",
    "#identifying variants significant after LDSR intercept correction\n",
    "diamante18_gws_LDSR <- subset(diamante18_gws, diamante18_gws$intercept_adj_p_values <= 0.00000005)\n",
    "diamante22_gws_LDSR <- subset(diamante22_gws, diamante22_gws$intercept_adj_p_values <= 0.00000005)\n",
    "\n",
    "#running simulation\n",
    "results_diamante22_t2dggi_LDSR <- calculate_replication(diamante22_gws_LDSR, t2dggi_gws)\n",
    "print(results_diamante22_t2dggi_LDSR)\n",
    "\n",
    "results_diamante18_t2dggi_LDSR <- calculate_replication(diamante18_gws_LDSR, t2dggi_gws)\n",
    "print(results_diamante18_t2dggi_LDSR)\n",
    "\n",
    "results_diamante18_diamante22_LDSR <- calculate_replication(diamante18_gws_LDSR, diamante22_gws)\n",
    "print(results_diamante18_diamante22_LDSR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41cee8f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "########################## Calculating replication proportions for simulations #################\n",
    "uncorr_diam18_diam22 <- results_diamante18_diamante22\n",
    "uncorr_diam18_t2dggi <- results_diamante18_t2dggi\n",
    "uncorr_diam22_t2dggi <- results_diamante22_t2dggi\n",
    "gc_diam22_t2dggi <- results_diamante22_t2dggi_GC\n",
    "ldsr_diam22_t2dggi <- results_diamante22_t2dggi_LDSR\n",
    "\n",
    "uncorr_diam18_diam22 <- uncorr_diam18_diam22 %>% rename(CHR = chromosome)\n",
    "uncorr_diam18_t2dggi <- uncorr_diam18_t2dggi %>% rename(CHR = chromosome)\n",
    "uncorr_diam22_t2dggi <- uncorr_diam22_t2dggi %>% rename(CHR = chromosome)\n",
    "gc_diam22_t2dggi <- gc_diam22_t2dggi %>% rename(CHR = chromosome)\n",
    "ldsr_diam22_t2dggi <- ldsr_diam22_t2dggi %>% rename(CHR = chromosome)\n",
    "\n",
    "gw_significant_counts_diamante_18 <- diamante18_gws %>%\n",
    "  group_by(CHR) %>%\n",
    "  summarise(total_gw_snps = n(), .groups = \"drop\")\n",
    "\n",
    "gw_significant_counts_diamante_22 <- diamante22_gws %>%\n",
    "  group_by(CHR) %>%\n",
    "  summarise(total_gw_snps = n(), .groups = \"drop\")\n",
    "\n",
    "uncorr_diam18_diam22 <- uncorr_diam18_diam22 %>%\n",
    "  left_join(gw_significant_counts_diamante_18, by = \"CHR\") %>%\n",
    "  mutate(simulation_replication_proportion = average_simulation / total_gw_snps)\n",
    "\n",
    "uncorr_diam18_t2dggi <- uncorr_diam18_t2dggi %>%\n",
    "  left_join(gw_significant_counts_diamante_18, by = \"CHR\") %>%\n",
    "  mutate(simulation_replication_proportion = average_simulation / total_gw_snps)\n",
    "\n",
    "uncorr_diam22_t2dggi <- uncorr_diam22_t2dggi %>%\n",
    "  left_join(gw_significant_counts_diamante_22, by = \"CHR\") %>%\n",
    "  mutate(simulation_replication_proportion = average_simulation / total_gw_snps)\n",
    "\n",
    "gc_diam22_t2dggi <- gc_diam22_t2dggi %>%\n",
    "  left_join(gw_significant_counts_diamante_22, by = \"CHR\") %>%\n",
    "  mutate(\n",
    "    simulation_replication_proportion = average_simulation / total_gw_snps,\n",
    "    real_replication_proportion = replicated_snps / total_gw_snps\n",
    "  )\n",
    "\n",
    "ldsr_diam22_t2dggi <- ldsr_diam22_t2dggi %>%\n",
    "  left_join(gw_significant_counts_diamante_22, by = \"CHR\") %>%\n",
    "  mutate(\n",
    "    simulation_replication_proportion = average_simulation / total_gw_snps,\n",
    "    real_replication_proportion = replicated_snps / total_gw_snps\n",
    "  )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
